src/compute-client/src/controller/sequential_hydration.rs

```diff
@@ -0,0 +1,262 @@
+    async fn absorb_command(&mut self, cmd: ComputeCommand<T>) -> Result<(), anyhow::Error> {
+        // Whether to forward this command to the wrapped client.
+        let mut forward = true;
+
+        match &cmd {
+            // We enforce sequential hydration only for non-transient dataflows, assuming that
+            // transient dataflows are created for interactive user queries and should always be
+            // scheduled as soon as possible.
+            ComputeCommand::CreateDataflow(dataflow) if !dataflow.is_transient() => {
+                let export_ids: Vec<_> = dataflow.export_ids().collect();
+                let id = export_ids.expect_element(|| "multi-export dataflows are not supported");
+                let as_of = dataflow.as_of.clone().unwrap();
+
+                debug!(%id, ?as_of, "tracking collection");
+                self.collections.insert(id, Collection::new(as_of));
+            }
+            ComputeCommand::Schedule(id) => {
+                if let Some(collection) = self.collections.get_mut(id) {
+                    debug!(%id, "enqueuing collection for hydration");
+                    self.hydration_queue.push_back(*id);
+                    collection.set_scheduled();
+                    forward = false;
+                }
+            }
+            ComputeCommand::AllowCompaction { id, frontier } if frontier.is_empty() => {
+                // The collection was dropped by the controller. Remove it from the tracking state
+                // to ensure we don't produce any more commands for it.
+                if self.collections.remove(id).is_some() {
+                    debug!(%id, "collection dropped");
+                }
+            }
+            _ => (),
+        }
+
+        if forward {
+            self.client.send(cmd).await?;
+        }
+
+        // Schedule collections that are ready now.
+        self.hydrate_collections().await
+    }
+
+    /// Observe a response and send resulting commands to the wrapped client.
+    async fn observe_response(&mut self, resp: &ComputeResponse<T>) -> Result<(), anyhow::Error> {
+        if let ComputeResponse::FrontierUpper { id, upper } = resp {
+            if let Some(collection) = self.collections.remove(id) {
+                let hydrated = PartialOrder::less_than(&collection.as_of, upper);
+                if hydrated || upper.is_empty() {
+                    debug!(%id, "collection hydrated");
+
+                    match collection.state {
+                        State::Created => {
+                            // We haven't seen a `Schedule` command yet, so no obligations to send
+                            // one either.
+                        }
+                        State::Scheduled => {
+                            // We are holding back the `Schedule` command for this collection. Send
+                            // it now.
+                            self.send(ComputeCommand::Schedule(*id)).await?;
```

Comment:
I'm not sure I follow this part. We've received a `FrontierUpper` for a collection, got a schedule instruction, but haven't yet told the replica to schedule the `id`. Doesn't this mean that the replica still scheduled the dataflow, because where else would the upper response come from that matches `hydrated`? Why do we still have to send the schedule command in this case?

Reply:
> Doesn't this mean that the replica still scheduled the dataflow

On the replica side, scheduling a dataflow means unsuspending it, which specifically means unsuspending its `persist_source`s. If we didn't forward the `Schedule` command here, the replica has not unsuspended the dataflow. But the dataflow can still hydrate if it doesn't read from any `persist_source`s, i.e. it only reads from indexes and/or constant inputs. It's true that in this case we don't have to send a `Schedule` command because it wouldn't have an effect, but I think it's still good to not just swallow them, to make the communication more predictable.

But apart from predictability there is actually a reason why we _have_ to send a `Schedule` command here: `REFRESH` MVs can jump their frontier ahead and will appear hydrated even if they read from persist sources and have not been scheduled yet. If we just dropped the `Schedule` command here, those `REFRESH` MVs would never be unsuspended and would become stuck.

Once we fix hydration status reporting for `REFRESH` MVs (https://github.com/MaterializeInc/materialize/issues/25518) the second reason won't apply anymore, but the predictability argument will still do so.

Reply:
That makes sense! Probably good to add a comment repeating what you wrote here.

Reply:
Done!